### PR TITLE
fix(env#namespace): only remove `lookups` that match full directory name

### PR DESCRIPTION
**Pull request by jamestalmage**
The existing method of removing sub-generator `lookup` patterns from the path created unexpected results in a few corner cases:

- `my-lib/generators/app` => `my-:app`, instead of `my-lib:app`
- `my.lib/generators/app` => `mylib:app`, instead of `my.lib:app`
- `test/generators/generators-thing` => `test:-thing`

### CODE REVIEW

File: lib/environment.js [hunk 1]
```diff
@@ -54,7 +54,7 @@ var Environment = module.exports = function Environment(args, opts, adapter) {
   this.runLoop.setMaxListeners(0);
   this.sharedFs.setMaxListeners(0);
 
-  this.lookups = ['.', 'generators', 'lib/generators'];
+  this.lookups = ['.', '..', 'generators', 'lib/generators'];
```
Comment (SBoudrias) [inline lib/environment.js]: Why adding `..` to the lookups dir?

Comment (jamestalmage) [inline lib/environment.js]: So `..` directories get removed.

I've just realized that value is also used in `resolve.js`, and I shouldn't have done that.

File: lib/environment.js [hunk 2]
```diff
@@ -391,7 +391,6 @@ Environment.prototype.namespace = function namespace(filepath) {
   ns = ns
     .replace(/(.*generator-)/, '') // remove before `generator-`
     .replace(/[\/\\](index|main)$/, '') // remove `/index` or `/main`
-    .replace(/\.+/g, '') // remove `.`
```
Comment (jamestalmage) [inline lib/environment.js]: It allowed removal of this line, which was causing problems if you wanted to name your package `generator-my.thing`. (requiring `yo mything` instead of `yo my.thing`)

File: lib/environment.js [hunk 3]
```diff
@@ -375,11 +375,11 @@ Environment.prototype.namespace = function namespace(filepath) {
   var ns = path.normalize(filepath.replace(new RegExp(escapeStrRe(path.extname(filepath)) + '$'), ''));
 
   // Sort lookups by length so biggest are removed first
-  var lookups = _(this.lookups).map(path.normalize).sortBy('length').value().reverse();
+  var lookups = _(this.lookups.concat('..')).map(path.normalize).sortBy('length').value().reverse();
```
Comment (jamestalmage) [inline lib/environment.js]: This alternate way should accomplish the same thing without interfering with the `lib/resolver.js`.

Comment (SBoudrias) [inline lib/environment.js]: I'm pretty sure this is not doing what you want. If you pass a String, concat will handle the string as an array of characters. This mean it'll re-add the `.` in the lookup.

I'm not sure why your tests didn't fail this time.

Comment (jamestalmage) [inline lib/environment.js]: Just tested it in REPL:
```js
var a = ['foo', 'bar'];
a.concat('baz');
// [ 'foo', 'bar', 'baz' ]
```

Comment (jamestalmage) [inline lib/environment.js]: PS: tests only failed the first time due to exceeding a timeout (and only on one of the three travis runs). They all passed locally before every commit.

Comment (SBoudrias) [inline lib/environment.js]: Hum ok, still, make this an array so we rely on a clear behavior.

File: lib/environment.js [hunk 4]
```diff
@@ -375,11 +375,11 @@ Environment.prototype.namespace = function namespace(filepath) {
   var ns = path.normalize(filepath.replace(new RegExp(escapeStrRe(path.extname(filepath)) + '$'), ''));
 
   // Sort lookups by length so biggest are removed first
-  var lookups = _(this.lookups).map(path.normalize).sortBy('length').value().reverse();
+  var lookups = _(this.lookups.concat(['..'])).map(path.normalize).sortBy('length').value().reverse();
 
   // if `ns` contains a lookup dir in its path, remove it.
   ns = lookups.reduce(function (ns, lookup) {
-    return ns.replace(new RegExp(escapeStrRe(lookup), 'g'), '');
+    return ns.replace(new RegExp('(?:\\\\|/|^)' + escapeStrRe(lookup) + '(?=\\\\|/)', 'g'), '');
```
Comment (SBoudrias) [inline lib/environment.js]: Add a comment explaining what this regex is doing and why it's working.

**Comment by jamestalmage** [comment]
OK @SBoudrias, done.

squash?

**Comment by SBoudrias** [comment]
Yes please, LGTM. I'll merge once squashed.

**Comment by jamestalmage** [comment]
squashed

**Comment by SBoudrias** [comment]
Awesome, thanks a lot!
